File: apps/mobile/lib/src/features/reading_sync/data/reading_sync_api_client.dart
import 'dart:convert';

import 'package:dio/dio.dart';

import '../../../core/network/api_client.dart';
import '../../../core/network/api_envelope.dart';
import '../../../core/network/api_error.dart';

class ReadingSyncApiClient {
  ReadingSyncApiClient({Dio? dio, ApiClient? apiClient})
    : _dio = dio ?? apiClient?.dio ?? ApiClient().dio;

  final Dio _dio;

  Future<void> upsertBookMetadata({
    required String bookFingerprint,
    required String title,
    String? author,
    required String fileType,
    required String sourceLang,
    required String targetLang,
    required int chapterCount,
  }) {
    final requestData = <String, Object?>{
      'title': title,
      'fileType': fileType,
      'sourceLang': sourceLang,
      'targetLang': targetLang,
      'chapterCount': chapterCount,
    };
    if (author != null) {
      requestData['author'] = author;
    }
    return _request(
      'PUT',
      '/api/v1/reading/books/$bookFingerprint',
      data: requestData,
    );
  }

  Future<void> updateReadingProgress({
    required String bookFingerprint,
    required int currentChapterIndex,
    required int currentParagraphIndex,
    required int currentCharOffset,
    DateTime? lastReadAt,
  }) {
    final requestData = <String, Object?>{
      'currentChapterIndex': currentChapterIndex,
      'currentParagraphIndex': currentParagraphIndex,
      'currentCharOffset': currentCharOffset,
    };
    if (lastReadAt != null) {
      requestData['lastReadAt'] = lastReadAt.toUtc().toIso8601String();
    }
    return _request(
      'PATCH',
      '/api/v1/reading/books/$bookFingerprint/progress',
      data: requestData,
    );
  }

  Future<void> _request(
    String method,
    String path, {
    required Map<String, Object?> data,
  }) async {
    final response = await _dio.request<Object?>(
      path,
      data: data,
      options: Options(method: method, validateStatus: (_) => true),
    );
    final envelope = ApiEnvelope.fromJson(_decode(response.data));
    if (!envelope.success) {
      throw envelope.error ??
          const ApiError(code: 'UNKNOWN_ERROR', message: 'Request failed');
    }
  }

  Object? _decode(Object? data) {
    if (data is String) {
      return jsonDecode(data);
    }
    return data;
  }
}

File: apps/web-reader/services/readingSyncApiClient.ts
import { ofetch } from 'ofetch'

import type { ApiEnvelope } from '../types/api'
import { WebApiError } from '../types/api'
import type { WebFileType } from '../types/localData'

type Fetcher = (url: string, options: Record<string, unknown>) => Promise<unknown>

export interface ReadingSyncApiClientOptions {
  baseUrl?: string
  fetcher?: Fetcher
}

export interface UpsertBookMetadataRequest {
  bookFingerprint: string
  title: string
  author?: string
  fileType: WebFileType
  sourceLang: string
  targetLang: string
  chapterCount: number
}

export interface UpdateReadingProgressRequest {
  bookFingerprint: string
  currentChapterIndex: number
  currentParagraphIndex: number
  currentCharOffset: number
  lastReadAt?: string | null
}

export interface ReadingSyncApiClient {
  upsertBookMetadata(input: UpsertBookMetadataRequest): Promise<unknown>
  updateReadingProgress(input: UpdateReadingProgressRequest): Promise<unknown>
}

export function createReadingSyncApiClient(options: ReadingSyncApiClientOptions = {}): ReadingSyncApiClient {
  const baseUrl = options.baseUrl ?? ''
  const fetcher = options.fetcher ?? ofetch
  return {
    upsertBookMetadata(input) {
      return request(fetcher, `${baseUrl}/api/v1/reading/books/${input.bookFingerprint}`, {
        method: 'PUT',
        body: {
          title: input.title,
          ...(input.author ? { author: input.author } : {}),
          fileType: input.fileType,
          sourceLang: input.sourceLang,
          targetLang: input.targetLang,
          chapterCount: input.chapterCount,
        },
      })
    },
    updateReadingProgress(input) {
      return request(fetcher, `${baseUrl}/api/v1/reading/books/${input.bookFingerprint}/progress`, {
        method: 'PATCH',
        body: {
          currentChapterIndex: input.currentChapterIndex,
          currentParagraphIndex: input.currentParagraphIndex,
          currentCharOffset: input.currentCharOffset,
          ...(input.lastReadAt ? { lastReadAt: input.lastReadAt } : {}),
        },
      })
    },
  }
}

async function request(
  fetcher: Fetcher,
  url: string,
  options: Record<string, unknown>,
): Promise<unknown> {
  const envelope = await fetcher(url, options) as ApiEnvelope<unknown>
  if (!envelope.success) {
    const error = envelope.error ?? { code: 'UNKNOWN_ERROR', message: 'Request failed' }
    throw new WebApiError(error.code, error.message)
  }
  return envelope.data
}
